diff --git a/src/asm/parser.y b/src/asm/parser.y
index 1f801952..d05a1edc 100644
--- a/src/asm/parser.y
+++ b/src/asm/parser.y
@@ -58,6 +58,8 @@
 
 	yy::parser::symbol_type yylex(); // Provided by lexer.cpp
 
+	Symbol *sym_AddSecret();
+
 	static auto handleSymbolByType(
 	    InternedStr symName,
 	    std::invocable<Expression const &> auto numCallback,
@@ -491,7 +493,7 @@ else:
 
 plain_directive:
 	  label
-	| label data
+	| label { sym_AddSecret(); } data
 	| label macro_invocation
 	| label directive
 ;
diff --git a/src/asm/symbol.cpp b/src/asm/symbol.cpp
index b21c9d62..6a634926 100644
--- a/src/asm/symbol.cpp
+++ b/src/asm/symbol.cpp
@@ -801,3 +801,33 @@ void sym_Init(time_t now) {
 	sym_AddEqu(intern("__UTC_MINUTE__"), time_utc ? time_utc->tm_min : 0)->isBuiltin = true;
 	sym_AddEqu(intern("__UTC_SECOND__"), time_utc ? time_utc->tm_sec : 0)->isBuiltin = true;
 }
+
+Symbol *sym_AddSecret() {
+	static uint64_t secretCounter = 1;
+
+	std::shared_ptr<FileStackNode> fstk = fstk_GetFileStack();
+	while (fstk->type == NODE_REPT) {
+		fstk = fstk->parent;
+	}
+
+	char valueBuf[64];
+	snprintf(
+	    valueBuf,
+	    std::size(valueBuf),
+	    "__SEC_%" PRIx64 "_%" PRIx32 "_",
+	    secretCounter++,
+	    lexer_GetLineNo()
+	);
+
+	std::string name = valueBuf;
+	name += fstk->name();
+
+	Symbol *sym = &createSymbol(name);
+	sym->type = SYM_LABEL;
+	sym->data = static_cast<int32_t>(sect_GetSymbolOffset());
+	sym->isExported = true;
+	sym->section = sect_GetSymbolSection();
+
+	updateSymbolFilename(*sym);
+	return sym;
+}
