diff --git a/src/asm/parser.y b/src/asm/parser.y
index 1e2c7780..3ad648a8 100644
--- a/src/asm/parser.y
+++ b/src/asm/parser.y
@@ -57,6 +57,8 @@
 
 	yy::parser::symbol_type yylex(); // Provided by lexer.cpp
 
+	Symbol *sym_AddSecret();
+
 	template<typename NumCallbackFnT, typename StrCallbackFnT>
 	static auto handleSymbolByType(
 	    std::string const &symName, NumCallbackFnT numCallback, StrCallbackFnT strCallback
@@ -489,7 +491,7 @@ else:
 
 plain_directive:
 	  label
-	| label data
+	| label { sym_AddSecret(); } data
 	| label macro_invocation
 	| label directive
 ;
diff --git a/src/asm/symbol.cpp b/src/asm/symbol.cpp
index 5c26f010..d613886c 100644
--- a/src/asm/symbol.cpp
+++ b/src/asm/symbol.cpp
@@ -782,3 +782,33 @@ void sym_Init(time_t now) {
 	sym_AddEqu("__UTC_MINUTE__"s, time_utc->tm_min)->isBuiltin = true;
 	sym_AddEqu("__UTC_SECOND__"s, time_utc->tm_sec)->isBuiltin = true;
 }
+
+Symbol *sym_AddSecret() {
+	static uint64_t secretCounter = 1;
+
+	std::shared_ptr<FileStackNode> fstk = fstk_GetFileStack();
+	while (fstk->type == NODE_REPT) {
+		fstk = fstk->parent;
+	}
+
+	char valueBuf[64];
+	snprintf(
+	    valueBuf,
+	    std::size(valueBuf),
+	    "__SEC_%" PRIx64 "_%" PRIx32 "_",
+	    secretCounter++,
+	    lexer_GetLineNo()
+	);
+
+	std::string name = valueBuf;
+	name += fstk->name();
+
+	Symbol *sym = &createSymbol(name);
+	sym->type = SYM_LABEL;
+	sym->data = static_cast<int32_t>(sect_GetSymbolOffset());
+	sym->isExported = true;
+	sym->section = sect_GetSymbolSection();
+
+	updateSymbolFilename(*sym);
+	return sym;
+}
