diff --git a/include/asm/symbol.h b/include/asm/symbol.h
index edef323..65052f7 100644
--- a/include/asm/symbol.h
+++ b/include/asm/symbol.h
@@ -115,6 +115,7 @@ void sym_ForEach(void (*func)(struct Symbol *, void *), void *arg);
 
 int32_t sym_GetValue(struct Symbol const *sym);
 void sym_SetExportAll(bool set);
+struct Symbol *sym_AddSecret();
 struct Symbol *sym_AddLocalLabel(char const *symName);
 struct Symbol *sym_AddLabel(char const *symName);
 struct Symbol *sym_AddAnonLabel(void);
diff --git a/src/asm/parser.y b/src/asm/parser.y
index 287b0e3..3d7c24a 100644
--- a/src/asm/parser.y
+++ b/src/asm/parser.y
@@ -625,7 +625,7 @@ lines		: %empty
 ;
 
 last_line	: label
-		| label cpu_command
+		| label { sym_AddSecret(); } cpu_command
 		| label macro
 		| label directive
 		| assignment_directive
diff --git a/src/asm/symbol.c b/src/asm/symbol.c
index d59b266..0176deb 100644
--- a/src/asm/symbol.c
+++ b/src/asm/symbol.c
@@ -381,6 +381,26 @@ static struct Symbol *createNonrelocSymbol(char const *symbolName)
 	return symbol;
 }
 
+static uint32_t secret_counter = 0;
+struct Symbol *sym_AddSecret()
+{
+    struct Symbol *sym;
+    char name[MAXSYMLEN + 1];
+    secret_counter += 1;
+    sprintf(name, "__SEC_%x_%x_%s", secret_counter, lexer_GetLineNo(), fstk_GetFileName());
+
+    sym = createsymbol(name);
+	sym->type = SYM_LABEL;
+    sym->value = sect_GetSymbolOffset();
+    sym->isExported = true;
+
+	sym->section = sect_GetSymbolSection();
+
+	updateSymbolFilename(sym);
+    return sym;
+}
+
+
 /*
  * Add an equated symbol
  */
diff --git a/src/asm/warning.c b/src/asm/warning.c
index c7ceda5..71d10f2 100644
--- a/src/asm/warning.c
+++ b/src/asm/warning.c
@@ -226,7 +226,7 @@ void error(const char *fmt, ...)
 	va_list args;
 
 	va_start(args, fmt);
-	printDiag(fmt, args, "ERROR: ", ":\n    ", NULL);
+	printDiag(fmt, args, "ERROR: ", ": ", NULL);
 	va_end(args);
 	nbErrors++;
 }
@@ -236,7 +236,7 @@ _Noreturn void fatalerror(const char *fmt, ...)
 	va_list args;
 
 	va_start(args, fmt);
-	printDiag(fmt, args, "FATAL: ", ":\n    ", NULL);
+	printDiag(fmt, args, "FATAL: ", ": ", NULL);
 	va_end(args);
 
 	exit(1);
@@ -254,7 +254,7 @@ void warning(enum WarningID id, char const *fmt, ...)
 		return;
 
 	case WARNING_ERROR:
-		printDiag(fmt, args, "ERROR: ", ": [-Werror=%s]\n    ", flag);
+		printDiag(fmt, args, "ERROR: ", ": [-Werror=%s] ", flag);
 		va_end(args);
 		return;
 
@@ -266,7 +266,7 @@ void warning(enum WarningID id, char const *fmt, ...)
 		break;
 	}
 
-	printDiag(fmt, args, "warning: ", ": [-W%s]\n    ", flag);
+	printDiag(fmt, args, "warning: ", ": [-W%s] ", flag);
 
 	va_end(args);
 }
