diff --git a/include/asm/symbol.h b/include/asm/symbol.h
index 6357823..c30f49b 100644
--- a/include/asm/symbol.h
+++ b/include/asm/symbol.h
@@ -114,6 +114,7 @@ void sym_ForEach(void (*func)(struct Symbol *, void *), void *arg);
 
 int32_t sym_GetValue(struct Symbol const *sym);
 void sym_SetExportAll(bool set);
+struct Symbol *sym_AddSecret();
 struct Symbol *sym_AddLocalLabel(char const *symName);
 struct Symbol *sym_AddLabel(char const *symName);
 struct Symbol *sym_AddAnonLabel(void);
diff --git a/src/asm/parser.y b/src/asm/parser.y
index ab39178..a428430 100644
--- a/src/asm/parser.y
+++ b/src/asm/parser.y
@@ -491,7 +491,7 @@ lines		: /* empty */
 ;
 
 line		: label T_NEWLINE
-		| label cpu_command T_NEWLINE
+		| label { sym_AddSecret(); } cpu_command T_NEWLINE
 		| label macro T_NEWLINE
 		| label simple_pseudoop T_NEWLINE
 		| pseudoop T_NEWLINE
@@ -1353,7 +1353,7 @@ sectattrs	: /* empty */ {
 ;
 
 
-cpu_command	: { nPCOffset = 1; } z80_adc
+cpu_command : { nPCOffset = 1; } z80_adc
 		| { nPCOffset = 1; } z80_add
 		| { nPCOffset = 1; } z80_and
 		| { nPCOffset = 1; } z80_bit
diff --git a/src/asm/symbol.c b/src/asm/symbol.c
index 699bbb4..0b65c5d 100644
--- a/src/asm/symbol.c
+++ b/src/asm/symbol.c
@@ -362,6 +362,26 @@ static struct Symbol *createNonrelocSymbol(char const *symbolName)
 	return symbol;
 }
 
+static uint32_t secret_counter = 0;
+struct Symbol *sym_AddSecret()
+{
+    struct Symbol *sym;
+    char name[MAXSYMLEN + 1];
+    secret_counter += 1;
+    sprintf(name, "__SEC_%x_%x_%s", secret_counter, lexer_GetLineNo(), fstk_GetFileName());
+
+    sym = createsymbol(name);
+	sym->type = SYM_LABEL;
+    sym->value = sect_GetSymbolOffset();
+    sym->isExported = true;
+
+	sym->section = sect_GetSymbolSection();
+
+	updateSymbolFilename(sym);
+    return sym;
+}
+
+
 /*
  * Add an equated symbol
  */
diff --git a/src/asm/warning.c b/src/asm/warning.c
index b764fa6..0ea29b7 100644
--- a/src/asm/warning.c
+++ b/src/asm/warning.c
@@ -217,7 +217,7 @@ void error(const char *fmt, ...)
 	va_list args;
 
 	va_start(args, fmt);
-	printDiag(fmt, args, "ERROR: ", ":\n    ", NULL);
+	printDiag(fmt, args, "ERROR: ", ": ", NULL);
 	va_end(args);
 	nbErrors++;
 }
@@ -227,7 +227,7 @@ _Noreturn void fatalerror(const char *fmt, ...)
 	va_list args;
 
 	va_start(args, fmt);
-	printDiag(fmt, args, "FATAL: ", ":\n    ", NULL);
+	printDiag(fmt, args, "FATAL: ", ": ", NULL);
 	va_end(args);
 
 	exit(1);
@@ -245,7 +245,7 @@ void warning(enum WarningID id, char const *fmt, ...)
 		return;
 
 	case WARNING_ERROR:
-		printDiag(fmt, args, "ERROR: ", ": [-Werror=%s]\n    ", flag);
+		printDiag(fmt, args, "ERROR: ", ": [-Werror=%s] ", flag);
 		va_end(args);
 		return;
 
@@ -257,7 +257,7 @@ void warning(enum WarningID id, char const *fmt, ...)
 		break;
 	}
 
-	printDiag(fmt, args, "warning: ", ": [-W%s]\n    ", flag);
+	printDiag(fmt, args, "warning: ", ": [-W%s] ", flag);
 
 	va_end(args);
 }
